x-pack/elastic-agent/pkg/composable/providers/kubernetes/pod.go

```diff
@@ -239,3 +371,134 @@ func getEphemeralContainers(pod *kubernetes.Pod) ([]kubernetes.Container, []kube
+func newNamespacePodUpdater(handler podUpdaterHandlerFunc, store podUpdaterStore, locker sync.Locker) *namespacePodUpdater {
+	return &namespacePodUpdater{
+		handler: handler,
+		store:   store,
+		locker:  locker,
+	}
+}
+
+// OnUpdate handles update events on namespaces.
+func (n *namespacePodUpdater) OnUpdate(obj interface{}) {
+	ns, ok := obj.(*kubernetes.Namespace)
+	if !ok {
+		return
+	}
+
+	// n.store.List() returns a snapshot at this point. If a delete is received
+	// from the main watcher, this loop may generate an update event after the
+	// delete is processed, leaving configurations that would never be deleted.
+	// Also this loop can miss updates, what could leave outdated configurations.
+	// Avoid these issues by locking the processing of events from the main watcher.
+	if n.locker != nil {
+		n.locker.Lock()
+		defer n.locker.Unlock()
+	}
+	for _, pod := range n.store.List() {
+		pod, ok := pod.(*kubernetes.Pod)
+		if ok && pod.Namespace == ns.Name {
+			n.handler(pod)
+		}
+	}
+}
+
+// OnAdd handles add events on namespaces. Nothing to do, if pods are added to this
+// namespace they will generate their own add events.
+func (*namespacePodUpdater) OnAdd(interface{}) {}
+
+// OnDelete handles delete events on namespaces. Nothing to do, if pods are deleted from this
+// namespace they will generate their own delete events.
+func (*namespacePodUpdater) OnDelete(interface{}) {}
+
+// podTerminating returns true if a pod is marked for deletion or is in a phase beyond running.
+func podTerminating(pod *kubernetes.Pod) bool {
+	if pod.GetObjectMeta().GetDeletionTimestamp() != nil {
+		return true
+	}
+
+	switch pod.Status.Phase {
+	case kubernetes.PodRunning, kubernetes.PodPending:
+	default:
+		return true
+	}
+
+	return false
+}
+
+// podTerminated returns true if a pod is terminated, this method considers a
+// pod as terminated if none of its containers are running (or going to be running).
+func podTerminated(pod *kubernetes.Pod, containers []*containerInPod) bool {
+	// Pod is not marked for termination, so it is not terminated.
+	if !podTerminating(pod) {
+		return false
+	}
+
+	// If any container is running, the pod is not terminated yet.
+	for _, container := range containers {
+		if container.status.State.Running != nil {
+			return false
+		}
+	}
+
+	return true
+}
+
+// getContainersInPod returns all the containers defined in a pod and their statuses.
+// It includes init and ephemeral containers.
+func getContainersInPod(pod *kubernetes.Pod) []*containerInPod {
+	var containers []*containerInPod
+	for _, c := range pod.Spec.Containers {
+		containers = append(containers, &containerInPod{spec: c})
+	}
+	for _, c := range pod.Spec.InitContainers {
+		containers = append(containers, &containerInPod{spec: c})
+	}
+	for _, c := range pod.Spec.EphemeralContainers {
+		c := kubernetes.Container(c.EphemeralContainerCommon)
+		containers = append(containers, &containerInPod{spec: c})
+	}
```

Comment:
There are two places now where we collect all the containers in the pod and their statuses (the other [here](https://github.com/elastic/beats/blob/848c275cafb1b1ac8a26f58bb6eb0f266cadc3d3/x-pack/elastic-agent/pkg/composable/providers/kubernetes/pod.go#L139)), this can be error-prone in future refactors. I wonder if this could be unified.

Reply:
Fair point. I will tune it! :)